crates/pera-core/src/dwallet_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,483 @@
+        for (message, session_id) in messages {
+            // Update the manager with the new network decryption key share (if relevant).
+            let session = self
+                .mpc_sessions
+                .get(&session_id)
+                .ok_or(DwalletMPCError::MPCSessionNotFound { session_id })?;
+            if let MPCSessionStatus::Finished(public_output, private_output) =
+                session.status.clone()
+            {
+                self.update_dwallet_mpc_network_key(
+                    &session.session_info,
+                    public_output,
+                    private_output,
+                )?;
+            }
+
+            self.consensus_adapter
+                .submit_to_consensus(&vec![message], &self.epoch_store()?)
+                .await?;
+        }
+        Ok(())
+    }
+
+    /// Update the encryption of decryption key share with the new shares.
+    /// This function is called when the network DKG protocol is done.
+    fn update_dwallet_mpc_network_key(
+        &self,
+        session_info: &SessionInfo,
+        public_output: MPCPublicOutput,
+        private_output: MPCPrivateOutput,
+    ) -> DwalletMPCResult<()> {
+        if let MPCRound::NetworkDkg(key_type) = session_info.mpc_round {
+            let epoch_store = self.epoch_store()?;
+            let network_keys = epoch_store
+                .dwallet_mpc_network_keys
+                .get()
+                .ok_or(DwalletMPCError::MissingDwalletMPCDecryptionKeyShares)?;
+
+            network_keys.add_key_version(
+                epoch_store.clone(),
+                key_type,
+                private_output.clone(),
+                public_output,
+            )?;
+        }
+        Ok(())
+    }
+
+    fn epoch_store(&self) -> DwalletMPCResult<Arc<AuthorityPerEpochStore>> {
+        self.epoch_store
+            .upgrade()
+            .ok_or(DwalletMPCError::EpochEnded(self.epoch_id))
+    }
+
+    /// Handles a message by forwarding it to the relevant MPC session.
+    /// If the session does not exist, punish the sender.
+    pub(crate) fn handle_message(
+        &mut self,
+        message: &[u8],
+        authority_name: AuthorityName,
+        session_id: ObjectID,
+    ) -> DwalletMPCResult<()> {
+        if self.malicious_actors.contains(&authority_name) {
+            return Ok(());
+        }
+        let session = match self.mpc_sessions.get_mut(&session_id) {
+            Some(session) => session,
+            None => {
+                warn!(
+                    "received a message for an MPC session ID: `{:?}` which does not exist",
+                    session_id
+                );
+                self.malicious_actors.insert(authority_name);
+                return Ok(());
+            }
+        };
+        match session.handle_message(&DWalletMPCMessage {
+            message: message.to_vec(),
+            authority: authority_name,
+        }) {
```

Comment:
Assumption: `handle_message()` checks if a message from this sender is already registered, and punishes/flags as malicious if so. 

If this isn't correct - fix it. If it is, please add a line of documentation here that says it is this `handle_message()`s responsibility and not this function to check this behaviour.